lib/Infer/InstSynthesis.cpp

```diff
@@ -0,0 +1,1360 @@
+  Tmp.push_back(O);
+  // -> Compare inputs
+  for (auto const &In : I) {
+    unsigned Width = CompInstMap[In.first]->Width;
+    // with component inputs and the output
+    for (auto const &L_x : Tmp) {
+      if (Width == CompInstMap[L_x.first]->Width)
+        continue;
+      InvalidWirings.insert(std::make_pair(In.first, L_x.first));
+    }
+  }
+  // -> Compare outputs
+  for (auto const &L_y : R) {
+    unsigned Width = CompInstMap[L_y.first]->Width;
+    // with component inputs and the output
+    for (auto const &L_x : Tmp) {
+      // Don't constrain yourself
+      if (L_y.first.first == L_x.first.first)
+        continue;
+      if (Width == CompInstMap[L_x.first]->Width)
+        continue;
+      InvalidWirings.insert(std::make_pair(L_y.first, L_x.first));
+    }
+  }
+
+  // Don't wire a component's input to its output.
+  for (auto const &L_x : P)
+    for (auto const &L_y : R)
+      if (L_x.first.first == L_y.first.first)
+        InvalidWirings.insert(std::make_pair(L_x.first, L_y.first));
+  // Don't wire input to a component's output
+  for (auto const &L_x : I)
+    for (auto const &L_y : R)
+      InvalidWirings.insert(std::make_pair(L_x.first, L_y.first));
+  // Don't wire a component's input to the output
+  for (auto const &L_x : P)
+    InvalidWirings.insert(std::make_pair(L_x.first, O.first));
+  // Don't wire a component's input(s) to other component
+  // input(s) directly. The solver should decide through other
+  // wirings if such connections are possible. For example, a solver
+  // could find out that 1_1 = 0_ and 1_2 = 0_1 => 1_1 = 1_2, making
+  // the explicit wiring test (1_1 = 1_2) redundant.
+  for (unsigned J = 0; J < P.size(); ++J)
+    for (unsigned K = J+1; K < P.size(); ++K)
+      InvalidWirings.insert(std::make_pair(P[J].first, P[K].first));
+  // Similarly, don't wire an input with other inputs(s) explicitly
+  for (unsigned J = 0; J < I.size(); ++J)
+    for (unsigned K = J+1; K < I.size(); ++K)
+      InvalidWirings.insert(std::make_pair(I[J].first, I[K].first));
+}
+
+Inst *InstSynthesis::getConsistencyConstraint(InstContext &IC) {
+  Inst *Ret = IC.getConst(APInt(1, true));
+
+  if (DebugLevel > 2)
+    llvm::outs() << "consistency constraints:\n";
+  // Don't wire the outputs of two components
+  for (unsigned J = 0; J < R.size(); ++J) {
+    auto const &L_x = R[J];
+    for (unsigned K = J+1; K < R.size(); ++K) {
+      auto const &L_y = R[K];
+      InvalidWirings.insert(std::make_pair(L_x.first, L_y.first));
+      Inst *Ne = IC.getInst(Inst::Ne, 1, {L_x.second, L_y.second});
+      Ret = IC.getInst(Inst::And, 1, {Ret, Ne});
+      if (DebugLevel > 2)
+        llvm::outs() << getLocVarStr(L_x.first) << " != "
+                     << getLocVarStr(L_y.first) << "\n";
+    }
+  }
+
+  return Ret;
+}
+
+
+void InstSynthesis::addConstraints(std::vector<InstMapping> &PCs,
+                                   InstContext &IC) {
+  Inst *C = getConsistencyConstraint(IC);
```

Comment:
repeated use of C for different purposes is a bit odd here-- would the code look better if you just inlined the calls that are generating the constraints inside the emplace_back calls?